torch/_dynamo/compiled_autograd.py

```diff
@@ -45,26 +45,69 @@ def __init__(self, compiler_fn) -> None:
         self.proxy_mode = ProxyTorchDispatchMode(self.fx_tracer, "symbolic")
         self.hooks_proxy = None
 
-    def wrap_fake(self, x, source):
+    def wrap_fake(self, x, source, required_shape=None):
         assert isinstance(x, torch.Tensor)
+        if required_shape:
+            """
+            [Note: Required Shapes]
+
+            FSDP hooks have side effects where the sizes of parameters are be mutated in-place.
+            At the start of backwards, a parameter might have a sharded size of `N/world_size`.
+            It will then get dynamically resized to `N` in the hook.  Then resized back to
+            `N/world_size` at the end of backwards in another hook.
+
+            The autograd engine uniformly requires the parameters to be size `N`, and will throw
+            an error if it ever observes the size `N/world_size` sharded value.  The size
+            requirements are enforced via `input_metadata_` on each autograd node, which will store
+            the unsharded `N` size.
+
+            Rather than trying to model the metadata mutation side effects of hooks, we observe that
```

Comment:
What about for algorithms where this is not the case? If you have a user program where the metadata mutation is meaningful and does not cancel itself out, is this still sound?

Reply:
Well autograd will just error if the sizes don't match... so I don't think an algorithm that doesn't fix the sizes makes any sense.

Reply:
Right but if forward is compiled, wont aot_autograd run afoul of these same problems still?

Reply:
Note, there is no requirement here that they "cancel itself out".

The only requirement is your algorithm doesn't hit a shape error in eager autograd.  "If shape isn't X error" (the check in eager autograd) and "your program doesn't error" (our assumption) implies "shape is X".  We don't know what will cause shape to be X, but we assume something will make it happen.

Reply:
Thanks for explaining - I understand now.